cli: print range keys in `debug keys` and `debug range-data`

erikgrinaker opened the pull request:
**cli: print range keys in `debug range-data`**

This patch outputs MVCC range keys in `debug range-data`. These are
output at the end of each key span, following the usual order of
`rditer.IterateReplicaKeySpans()`.

Output for a point key (r41 range descriptor) followed by a range key
(MVCC range tombstone across table 40) is e.g.:

```
1659220060.993264767,0 /Local/Range/Table/40/RangeDescriptor (0x016b12b0000172647363001706bbdac3b49c7f09): [/Table/40, /Table/41)
	Raw:r41:/Table/4{0-1} [(n1,s1):1, next=2, gen=0]

1659220060.993264766,2147483647 /Table/4{0-1} (0xb000-0xb100): ""
```

Resolves #83583.

Release note: None

**storage: support range keys in `Engine.MVCCIterate()`**

This patch adds support for MVCC range keys in `MVCCIterate()`, allowing
the key type to be specified and passing overlapping point and/or range
keys to the callback.

Release note: None

**cli: support MVCC range keys in `debug keys`**

This patch displays MVCC range keys in `debug keys`, interleaved with
point keys. It also adds the `--type rangekeys` parameter to only
display range keys.

For example:

```
$ ./cockroach debug keys --type rangekeys ~/local/1/data
1659382586.284196453,2147483647 /Table/{3-5}:
1659382586.284196453,2147483647 /Table/{6-12}:
1659382586.284196453,2147483647 /Table/1{4-5}:
1659382586.284196453,2147483647 /Table/{16-21}:
1659382586.284196453,2147483647 /Table/2{2-3}:
1659382586.284196453,2147483647 /Table/{24-32}:
1659382586.284196453,2147483647 /Table/3{3-7}:
1659382586.284196453,2147483647 /Table/{38-47}:
1659382586.284196453,2147483647 /{Table/50-Max}:

$ ./cockroach debug keys ~/local/1/data
[...]
1659382586.576106063,0 /Table/11/1/47/1/2022-08-01T19:41:14.829118Z/1/0:
1659382591.516425402,0 /Table/11/1/100/1/2022-08-01T19:41:38.121076Z/1/0:
1659382586.284196453,2147483647 /Table/1{4-5}:
1659382586.284196453,2147483647 /Table/{16-21}:
1659382586.284196453,2147483647 /Table/2{2-3}:
1659382586.284196453,2147483647 /Table/{24-32}:
1659382586.627810749,0 /NamespaceTable/30/1/0/0/"defaultdb"/4/1:
1659382586.641281444,0 /NamespaceTable/30/1/0/0/"postgres"/4/1:
[...]
```

Release note: None

erikgrinaker committed:
cli: print range keys in `debug range-data`

This patch outputs MVCC range keys in `debug range-data`. These are
output at the end of each key span, following the usual order of
`rditer.IterateReplicaKeySpans()`.

Output for a point key (r41 range descriptor) followed by a range key
(MVCC range tombstone across table 40) is e.g.:

```
1659220060.993264767,0 /Local/Range/Table/40/RangeDescriptor (0x016b12b0000172647363001706bbdac3b49c7f09): [/Table/40, /Table/41)
	Raw:r41:/Table/4{0-1} [(n1,s1):1, next=2, gen=0]

1659220060.993264766,2147483647 /Table/4{0-1} (0xb000-0xb100): ""
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85387)
<!-- Reviewable:end -->

erikgrinaker committed:
storage: support range keys in `Engine.MVCCIterate()`

This patch adds support for MVCC range keys in `MVCCIterate()`, allowing
the key type to be specified and passing overlapping point and/or range
keys to the callback.

Release note: None

erikgrinaker committed:
cli: support MVCC range keys in `debug keys`

This patch displays MVCC range keys in `debug keys`, interleaved with
point keys. It also adds the `--type rangekeys` parameter to only
display range keys.

For example:

```
$ ./cockroach debug keys --type rangekeys ~/local/1/data
1659382586.284196453,2147483647 /Table/{3-5}:
1659382586.284196453,2147483647 /Table/{6-12}:
1659382586.284196453,2147483647 /Table/1{4-5}:
1659382586.284196453,2147483647 /Table/{16-21}:
1659382586.284196453,2147483647 /Table/2{2-3}:
1659382586.284196453,2147483647 /Table/{24-32}:
1659382586.284196453,2147483647 /Table/3{3-7}:
1659382586.284196453,2147483647 /Table/{38-47}:
1659382586.284196453,2147483647 /{Table/50-Max}:

$ ./cockroach debug keys ~/local/1/data
[...]
1659382586.576106063,0 /Table/11/1/47/1/2022-08-01T19:41:14.829118Z/1/0:
1659382591.516425402,0 /Table/11/1/100/1/2022-08-01T19:41:38.121076Z/1/0:
1659382586.284196453,2147483647 /Table/1{4-5}:
1659382586.284196453,2147483647 /Table/{16-21}:
1659382586.284196453,2147483647 /Table/2{2-3}:
1659382586.284196453,2147483647 /Table/{24-32}:
1659382586.627810749,0 /NamespaceTable/30/1/0/0/"defaultdb"/4/1:
1659382586.641281444,0 /NamespaceTable/30/1/0/0/"postgres"/4/1:
[...]
```

Release note: None

nicktrav reviewed (approved):
Nice. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1, 14 of 14 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85387#-:-N8QfitbBw39J1IVUGNa:b-i2pyn4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/debug_print.go` line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85387#-N8QeTwX1m9ZoxH8W8yh:-N8QeTwX1m9ZoxH8W8yi:b-m27vn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/58d66c7ff51d847188a57603369f6e6223782d31/pkg/kv/kvserver/debug_print.go#L152)):*
> ```Go
> // SprintMVCCRangeKeyValue is like PrintMVCCRangeKeyValue, but returns a string.
> // If printKey is true, prints the key and the value together; otherwise, prints
> // just the value. For now, it
> ```

nit: incomplete sentence?


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85387)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aliher1911, @nicktrav, and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/debug_print.go` line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85387#-N8QeTwX1m9ZoxH8W8yh:-N8Sn4xW7XIT-VHwNGaz:b-cl9ani) ([raw file](https://github.com/cockroachdb/cockroach/blob/58d66c7ff51d847188a57603369f6e6223782d31/pkg/kv/kvserver/debug_print.go#L152)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

nit: incomplete sentence?
</blockquote></details>

Whoops. I was going to take a shortcut, but changed my mind mid-sentence. :) Thanks!


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
TFTR!

bors r=nicktrav
